Fix activation functions support

dmonllao opened the pull request:
As mentioned in #142 sigmoid derivative is hardcoded into backpropagate function which blocks any other activation function from working at all.

Included changes:
- Backpropagation using the neuron activation functions derivative instead of hardcoded sigmoid derivative
- Added missing activation functions derivatives
- Sigmoid forced for the output layer
- Updated ThresholdedReLU default threshold to 0 (acts as a ReLU)
- Unit tests for derivatives
- Unit tests for classifiers using different activation functions
- Added missing docs

Extra comments:
- I've used the term differentiate (although derive sounds more natural to me as a non English native speaker heheh)
- @param float|int to align the new method with compute (although they both shoudl only accept floats)
- Changed the default threshold in ThresholdedReLU so by default it acts like a basic ReLU (no need to create an extra ReLU class and no threshold is the most popular configuration)
- sigmoid-like activation functions returning values from 0 to 1 should still be used for the output layer; softmax is preferred over sigmoid for multiclass for what I've read but it would require changing APIs significantly, which I would leave for the NN future major refactoring

I haven't checked if this patch conflicts with #160 (which should also be ready for integration)

dmonllao commented:
I almost forgot to mention, I have no idea about how to calculate gaussian function derivative. I would appreciate if someone can check the one I've coded (copied from a python implementation).

MustafaKarabulut commented:
> I almost forgot to mention, I have no idea about how to calculate gaussian function derivative. I would appreciate if someone can check the one I've coded (copied from a python implementation).
Hi @dmonllao,
The derivative is correct. 

You can always use an online calculator to check it: https://www.derivative-calculator.net/#expr=e%5E%28-x%5E2%29

alchimik commented on src/Phpml/NeuralNetwork/ActivationFunction/Sigmoid.php:
```diff
@@ -25,4 +25,12 @@ public function compute($value): float
+     */
+    public function differentiate($value): float
+    {
+        return $value * (1 - $value);
```
```php
$s = $this->compute($value); 
return $s * (1 - $s);
```

alchimik commented on src/Phpml/NeuralNetwork/ActivationFunction/HyperbolicTangent.php:
```diff
@@ -25,4 +25,12 @@ public function compute($value): float
+     */
+    public function differentiate($value): float
+    {
+        return 1 - pow($value, 2);
```
```php
return 1 - ($this->compute($value) ** 2);
```

dmonllao commented:
Hi @alchimic and thanks for your review.

I am afraid I don't understand your suggestions; differentiate() input $value is the activated value (e.g. the output of sigmoid() or tanh()) 

https://github.com/php-ai/php-ml/pull/163/files#diff-c946c6c707b2bbe0eed833e66b85ea18L62 -> https://github.com/php-ai/php-ml/pull/163/files#diff-5350d5dfbc34518ae3f1b099040606bdR65. $neuron->getOutput() contains the activated value.

alchimik commented:
If '$value' == activated value, then sigmoid and tanh is right, but gaussian is wrong.

dmonllao commented:
You are right alchimik, thanks for commenting; I based this proposal on APIs that do not include the linear value as an argument to calculate the derivative, probably because they do not have gaussian as an activation function or any other activation function that not only depends on f(x) but also on x.

I've changed getDerivative() signature and stored the calculated z in Neuron. I've modified differentiate signature to receive both the value and the calculated value as most activation functions can reuse the calculation (these neural nets are slow enough already)

I've updated the pull request. Thanks again. Waiting for travis, but this seems ready to merge now.

dmonllao commented:
is there something you would like me to update here @itcraftsmanpl?

itcraftsmanpl commented:
Actually no, you can rebase and we can merge this immediately :D

dmonllao commented:
I've just updated the branch and travis is passing in my repo (https://travis-ci.org/dmonllao/php-ml) the error reported above seems to be a network time out. I do not have powers to trigger the travis job again.

itcraftsmanpl commented:
Big thanks 👍
